# Ann John
# Matriculation No.50784111

# Part I: Data Manipulation and Analysis

### 1. Import the dataset into a pandas dataframe. Make sure that the date column is in pandas data time format

In [2]:
import pandas as pd

In [3]:
df = pd.read_csv('train.csv')
df['datetime'] = pd.to_datetime(df['datetime'])

In [4]:
df.head()

datetime  season  holiday  workingday  weather  temp   atemp  \
0 2011-01-01 00:00:00       1        0           0        1  9.84  14.395   
1 2011-01-01 01:00:00       1        0           0        1  9.02  13.635   
2 2011-01-01 02:00:00       1        0           0        1  9.02  13.635   
3 2011-01-01 03:00:00       1        0           0        1  9.84  14.395   
4 2011-01-01 04:00:00       1        0           0        1  9.84  14.395   

   humidity  windspeed  casual  registered  count  
0        81        0.0       3          13     16  
1        80        0.0       8          32     40  
2        80        0.0       5          27     32  
3        75        0.0       3          10     13  
4        75        0.0       0           1      1

### 2. Check the data type of each column. How many rows are there in the dataset ? Does the dataset contain any missing values ?

In [5]:
print(df.dtypes)

datetime      datetime64[ns]
season                 int64
holiday                int64
workingday             int64
weather                int64
temp                 float64
atemp                float64
humidity               int64
windspeed            float64
casual                 int64
registered             int64
count                  int64
dtype: object


In [6]:
df.shape[0]

10886

#### There are 10886 rows in the dataset.

In [7]:
print(df.isnull().sum())

datetime      0
season        0
holiday       0
workingday    0
weather       0
temp          0
atemp         0
humidity      0
windspeed     0
casual        0
registered    0
count         0
dtype: int64


#### There are no null values in the dataset

### 3.Using the date column, create new columns for: year, month, day of the week and hour of the day.

In [8]:
df['year'] = df['datetime'].dt.year
df['month'] = df['datetime'].dt.month
df['day_of_week'] = df['datetime'].dt.day_name()
df['hour'] = df['datetime'].dt.hour

### 4. Rename the values in the season column to spring, summer, fall and winter.

In [9]:
season_map = {1: 'spring', 2: 'summer', 3: 'fall', 4: 'winter'}
df['season'] = df['season'].map(season_map)

In [10]:
df

datetime  season  holiday  workingday  weather   temp  \
0     2011-01-01 00:00:00  spring        0           0        1   9.84   
1     2011-01-01 01:00:00  spring        0           0        1   9.02   
2     2011-01-01 02:00:00  spring        0           0        1   9.02   
3     2011-01-01 03:00:00  spring        0           0        1   9.84   
4     2011-01-01 04:00:00  spring        0           0        1   9.84   
...                   ...     ...      ...         ...      ...    ...   
10881 2012-12-19 19:00:00  winter        0           1        1  15.58   
10882 2012-12-19 20:00:00  winter        0           1        1  14.76   
10883 2012-12-19 21:00:00  winter        0           1        1  13.94   
10884 2012-12-19 22:00:00  winter        0           1        1  13.94   
10885 2012-12-19 23:00:00  winter        0           1        1  13.12   

        atemp  humidity  windspeed  casual  registered  count  year  month  \
0      14.395        81     0.0000       3          13     16  2011      1   
1      13.635        80     0.0000       8          32     40  2011      1   
2      13.635        80     0.0000       5          27     32  2011      1   
3      14.395        75     0.0000       3          10     13  2011      1   
4      14.395        75     0.0000       0           1      1  2011      1   
...       ...       ...        ...     ...         ...    ...   ...    ...   
10881  19.695        50    26.0027       7         329    336  2012     12   
10882  17.425        57    15.0013      10         231    241  2012     12   
10883  15.910        61    15.0013       4         164    168  2012     12   
10884  17.425        61     6.0032      12         117    129  2012     12   
10885  16.665        66     8.9981       4          84     88  2012     12   

      day_of_week  hour  
0        Saturday     0  
1        Saturday     1  
2        Saturday     2  
3        Saturday     3  
4        Saturday     4  
...           ...   ...  
10881   Wednesday    19  
10882   Wednesday    20  
10883   Wednesday    21  
10884   Wednesday    22  
10885   Wednesday    23  

[10886 rows x 16 columns]

### 5.Calculate the total number of casual and registered bikes rented in the years 2011 and 2012.

In [11]:
rentals_by_year = df.groupby('year')[['casual', 'registered']].sum()
print(rentals_by_year)

      casual  registered
year                    
2011  155817      626162
2012  236318     1067179


### 6.Calculate the mean of the hourly total rentals count by season. Which season has the highest mean ?

In [35]:
df.groupby('season')['count'].mean()

season
fall      234.417124
spring    116.343261
summer    215.251372
winter    198.988296
Name: count, dtype: float64

#### Fall is the season that has the highest mean

### 7.Are more bikes rented by registered users on working or non-working days ? Does the answer differ for non-registered users ? Is the answer the same for both years ?

In [13]:
working_day_rentals = df.groupby(['workingday']) [['registered']].sum()
print(working_day_rentals)

            registered
workingday            
0               448835
1              1244506


#### Bikes rented by registered users on working days are more than non-working days. 

In [14]:
working_day_rentals = df.groupby(['workingday']) [['casual']].sum()
print(working_day_rentals)

            casual
workingday        
0           206037
1           186098


#### Yes the the answer does differ for non-registered users

In [15]:
# Ensure datetime column is in datetime format
df['datetime'] = pd.to_datetime(df['datetime'])

# Extract year from the datetime column
df['year'] = df['datetime'].dt.year

# Sum registered and casual rentals grouped by year and workingday
registered_by_year = df.groupby(['year', 'workingday'])['registered'].sum()
casual_by_year = df.groupby(['year', 'workingday'])['casual'].sum()

# Display results
print("Registered Rentals by Year and Working Day:\n", registered_by_year)
print("Casual Rentals by Year and Working Day:\n", casual_by_year)


Registered Rentals by Year and Working Day:
 year  workingday
2011  0             167492
      1             458670
2012  0             281343
      1             785836
Name: registered, dtype: int64
Casual Rentals by Year and Working Day:
 year  workingday
2011  0              83099
      1              72718
2012  0             122938
      1             113380
Name: casual, dtype: int64


In [26]:
# Registered
#2011
(458670)/(167492+458670)

0.7325101171901202

In [27]:
#2012
(785836)/(281343+785836)

0.7363675634546781

In [28]:
# Casual 
# 2011
(72718)/(83099+72718)

0.46668848713555006

In [25]:
#2012
(113380)/(122938+113380)

0.47977724929967247

#### The results for both years for the registered users are not different, it is approximately 73% for both years. It is the same for the casual users approximately 47% in 2011 and 48% in 2012 (a slight difference).

### 8.Which months in the year 2011 have the highest and the lowest total number of bikes rented ? Repeat for the year 2012.

In [16]:
rentals_2011 = df[df['year'] == 2011].groupby('month')['count'].sum()
print(rentals_2011)

month
1     23552
2     32844
3     38735
4     50517
5     79713
6     89776
7     92848
8     83296
9     79104
10    79522
11    70889
12    61183
Name: count, dtype: int64


#### Highest total numbers of bike rented = July 
#### Lowest total numbers of bike rented = January 

In [17]:
rentals_2012 = df[df['year'] == 2012].groupby('month')['count'].sum()
print(rentals_2012)

month
1      56332
2      66269
3      94766
4     116885
5     120434
6     130957
7     121769
8     130220
9     133425
10    127912
11    105551
12     98977
Name: count, dtype: int64


#### Highest total numbers of bike rented = September 
#### Lowest total numbers of bike rented = January 

### 9.Which type of weather have the highest and lowest mean of the hourly total rentals count ?

In [18]:
mean_rentals_by_weather = df.groupby('weather')['count'].mean()
highest_weather = mean_rentals_by_weather.idxmax()
lowest_weather = mean_rentals_by_weather.idxmin()
print(mean_rentals_by_weather)
print(highest_weather)  
print(lowest_weather)

weather
1    205.236791
2    178.955540
3    118.846333
4    164.000000
Name: count, dtype: float64
1
3


#### Highest mean of the hourly total rentals count = weather type 1
#### lowest mean of the hourly total rentals count = weather type 3

### 10.Calculate the correlation between the hourly total rentals count and all the numerical columns in the dataset. Which column has the highest correlation with the total rentals count ?

In [19]:
import numpy as np

In [20]:
num=df.select_dtypes(include=np.number)

In [21]:
correlations = num.corr()

In [22]:
print(correlations['count'].sort_values(ascending=False))

count         1.000000
registered    0.970948
casual        0.690414
hour          0.400601
temp          0.394454
atemp         0.389784
year          0.260403
month         0.166862
windspeed     0.101369
workingday    0.011594
holiday      -0.005393
weather      -0.128655
humidity     -0.317371
Name: count, dtype: float64


#### The column with the highest correlation with the count (total rentals) is the registered column, with a correlation value of 0.970948. This indicates strong positive correlation.



### 11.Create a new categorical column called day_period, which can take four possible values: night, morning, afternoon and evening. These values correspond to the following binning of the hour column: 0-6: night, 6-12: morning, 12-6: afternoon, 6-24:evening.

In [23]:
def assign_period(hour):
    if 0 <= hour < 6:
        return 'night'
    elif 6 <= hour < 12:
        return 'morning'
    elif 12 <= hour < 18:
        return 'afternoon'
    else:
        return 'evening'

df['day_period'] = df['hour'].apply(assign_period)
df

datetime  season  holiday  workingday  weather   temp  \
0     2011-01-01 00:00:00  spring        0           0        1   9.84   
1     2011-01-01 01:00:00  spring        0           0        1   9.02   
2     2011-01-01 02:00:00  spring        0           0        1   9.02   
3     2011-01-01 03:00:00  spring        0           0        1   9.84   
4     2011-01-01 04:00:00  spring        0           0        1   9.84   
...                   ...     ...      ...         ...      ...    ...   
10881 2012-12-19 19:00:00  winter        0           1        1  15.58   
10882 2012-12-19 20:00:00  winter        0           1        1  14.76   
10883 2012-12-19 21:00:00  winter        0           1        1  13.94   
10884 2012-12-19 22:00:00  winter        0           1        1  13.94   
10885 2012-12-19 23:00:00  winter        0           1        1  13.12   

        atemp  humidity  windspeed  casual  registered  count  year  month  \
0      14.395        81     0.0000       3          13     16  2011      1   
1      13.635        80     0.0000       8          32     40  2011      1   
2      13.635        80     0.0000       5          27     32  2011      1   
3      14.395        75     0.0000       3          10     13  2011      1   
4      14.395        75     0.0000       0           1      1  2011      1   
...       ...       ...        ...     ...         ...    ...   ...    ...   
10881  19.695        50    26.0027       7         329    336  2012     12   
10882  17.425        57    15.0013      10         231    241  2012     12   
10883  15.910        61    15.0013       4         164    168  2012     12   
10884  17.425        61     6.0032      12         117    129  2012     12   
10885  16.665        66     8.9981       4          84     88  2012     12   

      day_of_week  hour day_period  
0        Saturday     0      night  
1        Saturday     1      night  
2        Saturday     2      night  
3        Saturday     3      night  
4        Saturday     4      night  
...           ...   ...        ...  
10881   Wednesday    19    evening  
10882   Wednesday    20    evening  
10883   Wednesday    21    evening  
10884   Wednesday    22    evening  
10885   Wednesday    23    evening  

[10886 rows x 17 columns]

### 12.Generate a pivot table for the mean of the hourly total rentals count, with the index set to the day period and the column set to the working day column. What can you observe from the table ?

In [24]:
pivot_table = df.pivot_table(values='count', index='day_period', columns='workingday', aggfunc='mean')
print(pivot_table)

workingday           0           1
day_period                        
afternoon   371.022989  266.196141
evening     180.182759  251.054662
morning     157.772414  234.353763
night        44.052083   16.217582


#### Observations by Day Period:
#### Afternoon:
#### The highest mean rentals occur in the afternoon, especially on non-working days with an average of 371 rentals and For working days, the rentals are lower (266), likely due to people being at work during this time.
#### Evening:
#### On working days, the evening shows a high mean rental count of 251, second only to the afternoon. This suggests significant activity after work hours.On non-working days, the evening rentals are lower (180) compared to afternoons, indicating reduced demand as the day winds down.
#### Morning:
#### The rentals are higher on working days (234) compared to non-working days (157). This may reflect commuter usage on working days.
#### Night
#### Rentals are minimal at night for both working and non-working days, with the lowest mean rentals of 44 (non-working) and 16 (working).
